File: Source/AngularApplication/App/Services/ValidationService.ts
"use strict";

module ServiceRegister
{
    export class ValidationService
    {
        public static $inject = ["$http", "apiBaseUrl"];

        constructor(private $http: angular.IHttpService, private apiBaseUrl: string)
        {
        }

        public validateBusinessId(businessId: string, organizationId: string, allowDuplicates: boolean): angular.IPromise<BusinessIdentifierValidationResult>
        {
            return this.$http.post(this.apiBaseUrl + "serviceregister/businessid", this.createBusinessIdValidationObject(businessId, organizationId, allowDuplicates))
                .then((response: angular.IHttpPromiseCallbackArg<any>): BusinessIdentifierValidationResult =>
                {
                    return new BusinessIdentifierValidationResult(response.data.isValid, response.data.reasonForInvalidity);
                });
        }

        public validatePhoneNumber(phoneNumber: string): angular.IPromise<boolean>
        {
            return this.$http.post(this.apiBaseUrl + "serviceregister/phonenumber", "\"" + phoneNumber + "\"")
                .then((response: angular.IHttpPromiseCallbackArg<any>): boolean =>
                {
                    return response.data;
                });
        }

        public validateEmailAddress(emailAddress: string): angular.IPromise<boolean>
        {
            return this.$http.post(this.apiBaseUrl + "serviceregister/emailaddress", "\"" + emailAddress + "\"")
                .then((response: angular.IHttpPromiseCallbackArg<any>): boolean =>
                {
                    return response.data;
                });
        }

        public validateWebAddress(webAddress: string): angular.IPromise<boolean>
        {
            return this.$http.post(this.apiBaseUrl + "serviceregister/webaddress", "\"" + webAddress + "\"")
                .then((response: angular.IHttpPromiseCallbackArg<any>): boolean =>
                {
                    return response.data;
                });
        }

        public validatePostalCode(postalCode: string): angular.IPromise<boolean>
        {
            return this.$http.post(this.apiBaseUrl + "serviceregister/postalcode", "\"" + postalCode + "\"")
                .then((response: angular.IHttpPromiseCallbackArg<any>): boolean =>
                {
                    return response.data;
                });
        }

        public validatePostOfficeBoxPostalCode(postalCode: string): angular.IPromise<boolean>
        {
            return this.$http.post(this.apiBaseUrl + "serviceregister/postofficeboxpostalcode", "\"" + postalCode + "\"")
                .then((response: angular.IHttpPromiseCallbackArg<any>): boolean =>
                {
                    return response.data;
                });
        }

        private createBusinessIdValidationObject(businessId: string, organizationId: string, allowDuplicates: boolean): any
        {
            return {
                businessId: businessId,
                organizationId: organizationId,
                allowDuplicates: allowDuplicates
            }
        }
    }
}

File: Source/ServiceRegister.Application/Validation/ValidationService.cs
using System;
using Affecto.Identifiers;
using Affecto.Identifiers.Finnish;
using ServiceRegister.Application.Organization;

namespace ServiceRegister.Application.Validation
{
    internal class ValidationService : IValidationService
    {
        private readonly IOrganizationRepository organizationRepository;

        public ValidationService(IOrganizationRepository organizationRepository)
        {
            if (organizationRepository == null)
            {
                throw new ArgumentNullException("organizationRepository");
            }
            this.organizationRepository = organizationRepository;
        }

        public IBusinessIdentifierValidationResult ValidateUniqueBusinessIdentifier(string businessId, Guid? organizationId)
        {
            IBusinessIdentifierValidationResult result = ValidateBusinessIdentifier(businessId);
            if (result.IsValid && organizationRepository.HasOrganization(businessId, organizationId))
            {
                return new BusinessIdentifierValidationResult(false, InvalidBusinessIdentifierReason.AlreadyExists);
            }
            return result;
        }

        public IBusinessIdentifierValidationResult ValidateBusinessIdentifier(string businessId)
        {
            var specification = new BusinessIdentifierSpecification();
            bool isValid = specification.IsSatisfiedBy(businessId);
            string reasonForInvalidity = specification.GetReasonsForDissatisfactionSeparatedWithNewLine();
            return new BusinessIdentifierValidationResult(isValid, reasonForInvalidity);
        }

        public bool ValidatePhoneNumber(string phoneNumber)
        {
            var specification = new PhoneNumberSpecification();
            return specification.IsSatisfiedBy(phoneNumber);
        }

        public bool ValidateEmailAddress(string emailAddress)
        {
            var specification = new EmailAddressSpecification();
            return specification.IsSatisfiedBy(emailAddress);
        }

        public bool ValidateWebAddress(string webAddress)
        {
            var specification = new WebAddressSpecification();
            return specification.IsSatisfiedBy(webAddress);
        }

        public bool ValidatePostalCode(string postalCode)
        {
            var specification = new PostalCodeSpecification();
            return specification.IsSatisfiedBy(postalCode);
        }

        public bool ValidatePostOfficeBoxPostalCode(string postalCode)
        {
            var specification = new PostOfficeBoxPostalCodeSpecification();
            return specification.IsSatisfiedBy(postalCode);
        }
    }
}
